File: apps/web/src/types/real-estate-contact.ts
/**
 * Real Estate Contact Types
 * Types pour les contacts des transactions immobilières
 */

export enum ContactType {
  CLIENT = "client",
  REAL_ESTATE_BROKER = "real_estate_broker",
  MORTGAGE_BROKER = "mortgage_broker",
  NOTARY = "notary",
  INSPECTOR = "inspector",
  CONTRACTOR = "contractor",
  INSURANCE_BROKER = "insurance_broker",
  OTHER = "other",
}

export interface RealEstateContact {
  id: number;
  first_name: string;
  last_name: string;
  email?: string;
  phone?: string;
  company?: string;
  type: ContactType;
  user_id?: number;
  created_at: string;
  updated_at: string;
}

export interface RealEstateContactCreate {
  first_name: string;
  last_name: string;
  email?: string;
  phone?: string;
  company?: string;
  type: ContactType;
  user_id?: number;
}

export interface RealEstateContactUpdate {
  first_name?: string;
  last_name?: string;
  email?: string;
  phone?: string;
  company?: string;
  type?: ContactType;
}

export interface TransactionContact {
  transaction_id: number;
  contact_id: number;
  role: string;
  created_at: string;
  contact: RealEstateContact;
}

export interface TransactionContactCreate {
  /** ID du contact real_estate_contacts (utiliser depuis la section Contacts immobiliers) */
  contact_id?: number;
  /** ID du contact réseau reseau/contacts (utiliser depuis la section Réseau) */
  reseau_contact_id?: number;
  role: string;
}

export interface RealEstateContactListResponse {
  contacts: RealEstateContact[];
  total: number;
  skip: number;
  limit: number;
}

export interface TransactionContactListResponse {
  contacts: TransactionContact[];
  total: number;
}

// Rôles prédéfinis pour les transactions
export const TRANSACTION_ROLES = [
  "Vendeur",
  "Acheteur",
  "Courtier immobilier (vendeur)",
  "Courtier immobilier (acheteur)",
  "Notaire instrumentant",
  "Inspecteur en bâtiments",
  "Arpenteur-géomètre",
  "Conseiller hypothécaire",
  "Autre",
] as const;

export type TransactionRole = typeof TRANSACTION_ROLES[number];

File: backend/app/models/real_estate_contact.py
"""
Real Estate Contact Model
Modèle pour les contacts des transactions immobilières
"""

import enum
from sqlalchemy import Column, Integer, String, ForeignKey, DateTime, Enum as SQLEnum, Index
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.core.database import Base


class ContactType(str, enum.Enum):
    """Type de contact pour les transactions immobilières"""
    CLIENT = "client"
    REAL_ESTATE_BROKER = "real_estate_broker"
    MORTGAGE_BROKER = "mortgage_broker"
    NOTARY = "notary"
    INSPECTOR = "inspector"
    CONTRACTOR = "contractor"
    INSURANCE_BROKER = "insurance_broker"
    OTHER = "other"


class RealEstateContact(Base):
    """Contact model for real estate transactions"""
    __tablename__ = "real_estate_contacts"
    __table_args__ = (
        Index("idx_real_estate_contacts_email", "email"),
        Index("idx_real_estate_contacts_type", "type"),
        Index("idx_real_estate_contacts_user_id", "user_id"),
        Index("idx_real_estate_contacts_created_at", "created_at"),
    )

    id = Column(Integer, primary_key=True, index=True)
    first_name = Column(String(100), nullable=False, index=True, comment="Prénom")
    last_name = Column(String(100), nullable=False, index=True, comment="Nom de famille")
    email = Column(String(255), nullable=True, unique=True, index=True, comment="Adresse email")
    phone = Column(String(50), nullable=True, comment="Numéro de téléphone")
    company = Column(String(200), nullable=True, comment="Entreprise ou agence")
    type = Column(SQLEnum(ContactType), nullable=False, index=True, comment="Type de contact")
    
    # Lien optionnel avec User (plusieurs contacts par utilisateur)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False, index=True)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)
    
    # Relationships
    user = relationship("User", backref="real_estate_contact")
    transaction_roles = relationship("TransactionContact", back_populates="contact", cascade="all, delete-orphan")
    
    def __repr__(self) -> str:
        return f"<RealEstateContact(id={self.id}, name={self.first_name} {self.last_name}, type={self.type})>"
    
    @property
    def full_name(self) -> str:
        """Retourne le nom complet"""
        return f"{self.first_name} {self.last_name}".strip()
